### PR TITLE
NNEstimator: Add option to handle invalid data during fit and transform

### PR DESCRIPTION
**Pull request by hhbyyh**
Add option `handleInvalid` to process faulty data.

This is a prototype and does not target for the first release.

### CODE REVIEW

**Comment by hhbyyh** [comment]
cc @jenniew

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/common/FeatureLabelPreprocessing.scala [hunk 1]
```diff
@@ -37,13 +37,19 @@ class FeatureLabelPreprocessing[F, L, T: ClassTag](
...
+          case Some(l) =>
+            val labelTensor = labelStep(Iterator(l)).next()
+            Sample[T](featureTensor, labelTensor)
+          case None =>
+            Sample[T](featureTensor)
+        }
+      } catch {
+        case e: Exception =>
+          println(e.getMessage)
+          null
```
Comment (jenniew) [inline zoo/src/main/scala/com/intel/analytics/zoo/feature/common/FeatureLabelPreprocessing.scala]: If we choose to set handleInvalid as error, I think we need to throw the original exception not the NullPointerException which is confusing to the user.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNEstimator.scala [hunk 2]
```diff
@@ -538,8 +577,11 @@ class NNModel[T: ClassTag] private[zoo] (
               "unexpected batchResult dimension: " + batchResult.size().mkString(", "))
           }
         }
-        rowBatch.toIterator.zip(predictions).map { case (row, predict) =>
+
+        validRow.map(_._1).iterator.zip(predictions).map { case (row, predict) =>
           Row.fromSeq(row.toSeq ++ Seq(predict))
+        } ++ invalidRows.map(_._1).map { row =>
+          Row.fromSeq(row.toSeq ++ Seq(Array.empty))
```
Comment (hhbyyh) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNEstimator.scala]: Row.fromSeq(row.toSeq ++ Seq(outputToPrediction(Tensor(0))))

And update in NNClassifier.scala:
  protected override def outputToPrediction(output: Tensor[T]): Any = {
    if (output.isSameSizeAs(Tensor(0))) {
      Double.NaN
    } else {
      ev.toType[Double](output.max(1)._2.valueAt(1))
    }
  }

File: zoo/src/main/scala/com/intel/analytics/zoo/feature/common/FeatureLabelPreprocessing.scala (hunk 1 above)
Comment (jason-dai) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/common/FeatureLabelPreprocessing.scala]: Throw a meaningful exception

Comment (hhbyyh) [reply zoo/src/main/scala/com/intel/analytics/zoo/feature/common/FeatureLabelPreprocessing.scala]: `null` here is the place holder for the invalid data.

**Comment by jason-dai** [comment]
Why don't we simply to implement a Preprocessing to filer invalid images, which can be reused across NNEstimator, NNModel, etc.?

**Comment by hhbyyh** [comment]
I was trying to support the filtering, not only for image data, but for other data types during preprocessing as well.
I do feel this is not in the best state as it's requesting all the Preprocessing to return null for invalid records and not-null for valid records, but we didn't enforce the design in Preprocessing.

**Comment by hhbyyh** [comment]
Options are:
        'drop': invalid data are ignored during training (fit).
                During prediction (transform), rows containing invalid data will be dropped.
        'error': throw an error whenever an invalid data is met.

**Pull request closed without merging**